dot/types/inherents.go

```diff
@@ -64,6 +68,29 @@ func (d *InherentsData) SetInt64Inherent(key []byte, data uint64) error {
 	return nil
 }
 
+// SetStructInherent sets a struct inherent.
+func (d *InherentsData) SetStructInherent(key []byte, value interface{}) error {
+	if len(key) != 8 {
+		return errors.New("inherent key must be 8 bytes")
+	}
+
+	data, err := scale.Marshal(value)
+	if err != nil {
+		return err
+	}
+
+	venc, err := scale.Marshal(data)
+	if err != nil {
+		return err
+	}
```

Comment:
wait why is this marshaling twice?

Reply:
since @kishansagathiya mentioned this method works, i think it's actually correct: https://spec.polkadot.network/#defn-scale-dictionary

maybe double check with the spec team, but it says here the encoding of each (k,v) in the entire map is firstly encoded before concatentation, so if polkadot is interpreting the value as a byte slice instead of a struct this would be correct

Reply:
Ahh interesting, thanks for clarifying 💯  Kishan and I went a bit crazy figuring this out last week, without really knowing why it was suddenly working by encoding it twice. We should probably add a comment to why it's like that.

Reply:
Actually, re-reading the spec (and also PR-ing the slightly odd grammar):

1. We need to encode the tuple (`key`, `value`) as `encodedTuple` := Enc(`key`) || Enc(`value`)
2. Then we encode the map ((key: value)) as: Enc(size) || `encodedTuple` = Enc(1) || `encodedTuple`

Now our code is quite wrong I think since:

- the key gets lost (we only have `encodedTuple := Enc(value)`)
- the size is wrong, since we set the length of the encoded value instead of the map number of elements (`1` here)

IMO we should add support for maps in our scale package before fixing this.

Reply:
It's still quite strange they want it as an encoded byte slice though, why not have it as a struct? 🤔